index.bs

```diff
@@ -678,34 +880,87 @@ urlPrefix: https://w3c.github.io/FileAPI/#dfn-; type: dfn;
+				a user gesture event and the user must select the paste option from the native context menu that pops up
+				when write() is called from JS, otherwise, the promise will be rejected.
+
+				1. If |r| is false, then:
+				
+					1. [=Queue a global task=] on the [=permission task source=], given |realm|'s [=Realm/global object=], to [=reject=] |p| with {{"NotAllowedError"}} {{DOMException}} in |realm|.
+
+					1. Abort these steps.
+
+				1. [=Queue a global task=] on the [=clipboard task source=], given |realm|'s [=Realm/global object=], to perform the below steps:
+
+					1. Let |itemList| and |cleanItemList| be an empty [=sequence=]&lt;{{Blob}}&gt;.
+
+					1. For each |clipboardItem| in |data|:
+
+						1. For each |item| in |clipboardItem|'s [=ClipboardItem/clipboard item=]'s [=list of representations=]:
+
+							1. Let |p1| be the |item|'s [=representation/data=].
+
+							1. [=promise/React=] to |p1|:
 
-				1. If |r| is not "granted", then reject |p| with a "NotAllowedError" DOMException
+								1. If |p1| was fulfilled with value |v|, then:
 
-				1. Let |cleanItemList| be an empty sequence&lt;{{Blob}}>.
+									1. If |v| is a {{DOMString}}, then follow the below steps:
 
-				1. For each {{Blob}} |item| in |data|:
+										1. Let |v1| be the result of [=/encoding=] |v|.
 
-					1. Let |cleanItem| be a sanitized copy of |item|.
+										1. Let |blobData| be a {{Blob}} created using |v1| with its {{Blob/type}} set to |mimeType|, [=serialize a MIME type|serialized=].
 
-					1. If unable to create a sanitized copy, then reject |p|.
+										1. Add |blobData| to |itemList|.
 
-					1. Add |cleanItem| to |cleanItemList|.
+									1. If |v| is a {{Blob}}, then add |v| to |itemList|.
 
-				1. Replace the [=system clipboard data=] with |cleanItemList|.
+								1. If |p1| was rejected, then:
 
-				1. Resolve |p|.
+									1. [=Reject=] |p| with {{"NotAllowedError"}} {{DOMException}} in |realm|.
+
+									1. Return |p|.
+
+						1. For each |blob| in |itemList|:
+
+							1. Let |type| be the |blob|'s {{Blob/type}}.
+
+							1. If |type| is not in the [=mandatory data types=] list, then [=reject=] |p| with {{"NotAllowedError"}} {{DOMException}} in |realm| and abort these steps.
+
+							1. Let |cleanItem| be a sanitized copy of |blob|.
+
+								Issue: Add definition of sanitized copy.
+
+							1. If unable to create a sanitized copy, then follow the below steps:
+							
+								1. [=Reject=] |p| with {{"NotAllowedError"}} DOMException in |realm|.
+
+								1. Return |p|.
+
+							1. Append |cleanItem| to |cleanItemList|.
+
+						1. Let |option| be |clipboardItem|'s [=ClipboardItem/clipboard item=]'s [=presentation style=].
+
+						1. [=write blobs and option to the clipboard=] with |cleanItemList| and |option|.
+
+							Note: Writing multiple [=/clipboard item=]s to the clipboard is only supported on MacOS.
+							For other platforms, only the first [=/clipboard item=] should be written to the clipboard.
```

Comment:
Including a normative "should" inside a non-normative note is not good. Especially one that modifies the algorithm so extensively.

I'd suggest the following: up near the top of the algorithm, maybe as step 2, add something like:

> If |data|'s [=list/size=] is greater than 1, and the current operating system does not support multiple [=clipboard items=] on the [=system clipboard data=], then set |data| to « |data|[0] ».

Reply:
Done.